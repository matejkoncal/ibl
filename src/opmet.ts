import { MessageTypeState } from "./messageTypeSelector/messageTypeSelector";

export type WeatherReport = {
  queryType: string;
  reportTime: string;
  stationId: string;
  text: string;
};

export async function query(airports: string, countries: string, messageTypes: MessageTypeState): Promise<Record<string, WeatherReport[]>> {
  const response = await fetch("https://ogcie.iblsoft.com/ria/opmetquery", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      id: "123",
      method: "query",
      params: [
        {
          id: "briefing01",
          reportTypes: [messageTypes.metar ? "METAR" : "", messageTypes.sigmet ? "SIGMET" : "", messageTypes.taf ? "TAF" : ""].filter(Boolean),
          stations: airports.split(" ").filter(Boolean),
          countries: countries.split(" ").filter(Boolean),
        },
      ],
    }),
  });

  if (!response.ok) {
    throw new Error("Network response was not ok");
  }

  const parsedResponse = await response.json();

  if (parsedResponse.error) {
    throw new Error(parsedResponse.error.message);
  }

  return Object.groupBy(parsedResponse.result as WeatherReport[], ({ stationId }) => stationId) as Record<string, WeatherReport[]>;
}
