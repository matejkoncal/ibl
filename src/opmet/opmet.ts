import { MessageTypeState } from "../messageTypeSelector/messageTypeSelector";
import { TableSource } from "../resultTable/resultTable";

export type WeatherReport = {
  queryType: string;
  reportTime: string;
  stationId: string;
  text: string;
};

export async function query(airports: string, countries: string, messageTypes: MessageTypeState): Promise<TableSource> {
  const response = await fetch("https://ogcie.iblsoft.com/ria/opmetquery", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      id: "123",
      method: "query",
      params: [
        {
          id: "briefing01",
          reportTypes: [messageTypes.metar ? "METAR" : "", messageTypes.sigmet ? "SIGMET" : "", messageTypes.taf ? "TAF" : ""].filter(Boolean),
          stations: airports.split(" ").filter(Boolean),
          countries: countries.split(" ").filter(Boolean),
        },
      ],
    }),
  });

  if (!response.ok) {
    throw new Error("Network response was not ok");
  }

  const parsedResponse = await response.json();

  if (parsedResponse.error) {
    throw new Error(parsedResponse.error.message);
  }

  return Object.groupBy(parsedResponse.result as WeatherReport[], ({ stationId }) => stationId) as TableSource;
}
