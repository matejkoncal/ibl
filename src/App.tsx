import { Box, Button, CssBaseline, Stack, TextField } from "@mui/material";
import "./App.css";
import { MessageTypeSelector, MessageTypeState } from "./messageTypeSelector/messageTypeSelector";
import { useState } from "react";
import { ResultTable } from "./resultTable/resultTable";
import { ToastContainer, toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
import { WeatherReport, query } from "./opmet";

function App() {
  const [airports, setAirports] = useState("");
  const [countries, setCountries] = useState("");
  const [messageTypes, setMessageTypes] = useState<MessageTypeState>({ metar: false, sigmet: false, taf: false });
  const [tableContent, setTableContent] = useState<Record<string, WeatherReport[]>>({});

  const haveSomeLocation = airports || countries;
  const haveSomeMessageType = messageTypes.metar || messageTypes.sigmet || messageTypes.taf;

  const canQuery = haveSomeLocation && haveSomeMessageType;

  return (
    <Box padding="30px" justifyContent="center">
      <CssBaseline />
      <Stack width="100%" alignItems="center" spacing={2}>
        <Stack spacing={2} width="100%" maxWidth="sm">
          <MessageTypeSelector onChange={setMessageTypes} />
          <TextField value={airports} onChange={e => setAirports(e.target.value.toUpperCase())} label="Airports" placeholder="LZIB LKPR EGLL" />
          <TextField value={countries} onChange={e => setCountries(e.target.value.toUpperCase())} label="Countries" placeholder="SQ CZ" />
          <Button
            onClick={async () => {
              try {
                const result = await query(airports, countries, messageTypes);
                setTableContent(result);
              } catch (error) {
                toast.error((error as Error).message);
              }
            }}
            disabled={!canQuery}
            variant="contained"
          >
            Create Briefing
          </Button>
        </Stack>
        <ResultTable tableSource={tableContent} />
      </Stack>
      <ToastContainer />
    </Box>
  );
}

export default App;
