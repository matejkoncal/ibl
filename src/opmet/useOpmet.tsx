import { useCallback, useState } from "react";
import { query } from "./opmet";
import { MessageTypeState } from "../messageTypeSelector/messageTypeSelector";
import { TableSource } from "../resultTable/resultTable";

export const useOpmet = () => {
  const [loading, setLoading] = useState(false);
  const [tableContent, setTableContent] = useState<TableSource>({});
  const [error, setError] = useState<Error | undefined>();

  const loadOpmet = useCallback((airports: string, countries: string, messageTypes: MessageTypeState) => {
    setLoading(true);
    setError(undefined);
    query(airports, countries, messageTypes)
      .then(result => setTableContent(result))
      .catch(setError)
      .finally(() => setLoading(false));
  }, []);

  return [loadOpmet, tableContent, loading, error] as const;
};
