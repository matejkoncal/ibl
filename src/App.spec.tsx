import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "./App";
import "@testing-library/jest-dom";
import * as opmet from "./opmet/opmet";

describe("App", () => {
  it("should render", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  it("should have disabled button when no location is set", () => {
    render(<App />);
    expect(screen.getByRole("button")).toHaveProperty("disabled", true);
  });

  it("should have disabled button when no message type is set but location is set", async () => {
    render(<App />);
    const airportsInput = screen.getByLabelText("Airports");
    await userEvent.type(airportsInput, "LZIB");
    expect(screen.getByRole("button")).toHaveProperty("disabled", true);
  });

  it("should have enabled button when location and message type is set", async () => {
    render(<App />);
    const airportsInput = screen.getByLabelText("Airports");
    await userEvent.type(airportsInput, "LZIB");
    const metarCheckbox = screen.getByLabelText("METAR");
    await userEvent.click(metarCheckbox);
    expect(screen.getByRole("button")).toHaveProperty("disabled", false);
  });

  it("should call query function when button is clicked", async () => {
    const querySpy = jest.spyOn(opmet, "query");

    render(<App />);
    const airportsInput = screen.getByLabelText("Airports");
    await userEvent.type(airportsInput, "LZIB");
    const metarCheckbox = screen.getByLabelText("METAR");
    await userEvent.click(metarCheckbox);
    const button = screen.getByRole("button");
    await userEvent.click(button);

    expect(querySpy).toHaveBeenCalled();
  });

  it("should show error toast when query function throws error", async () => {
    jest.spyOn(opmet, "query").mockRejectedValue(new Error("Network error"));

    render(<App />);
    const airportsInput = screen.getByLabelText("Airports");
    await userEvent.type(airportsInput, "LZIB");
    const metarCheckbox = screen.getByLabelText("METAR");
    await userEvent.click(metarCheckbox);
    const button = screen.getByRole("button");
    await userEvent.click(button);

    expect(await screen.findByText("Network error")).toBeInTheDocument();
  });
});
