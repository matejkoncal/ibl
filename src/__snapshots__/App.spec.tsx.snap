// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render 1`] = `
<div>
  <div
    class="MuiBox-root css-x8tma9"
  >
    <div
      class="MuiStack-root css-2fgnmr-MuiStack-root"
    >
      <div
        class="MuiStack-root css-f1jtf-MuiStack-root"
      >
        <div
          class="MuiStack-root css-pl8gqh-MuiStack-root"
          data-testid="messageTypeSelector"
        >
          <label
            class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
              data-testid="metar-checkbox"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate="false"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
            >
              METAR
            </span>
          </label>
          <label
            class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
              data-testid="sigmer-checkbox"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate="false"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
            >
              SIGMET
            </span>
          </label>
          <label
            class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
              data-testid="taf-checkbox"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate="false"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
            >
              TAF
            </span>
          </label>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r0:"
            id=":r0:-label"
          >
            Airports
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              placeholder="LZIB LKPR EGLL"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Airports
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r1:"
            id=":r1:-label"
          >
            Countries
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r1:"
              placeholder="SQ CZ"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Countries
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          Create Briefing
        </button>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
      >
        <table
          aria-label="result table"
          class="MuiTable-root css-rqglhn-MuiTable-root"
        >
          <tbody
            class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          />
        </table>
      </div>
    </div>
    <div
      class="Toastify"
    />
  </div>
</div>
`;
